mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala

```diff
@@ -581,21 +609,33 @@ private[ml] object DefaultParamsReader {
   }
 
   /**
-   * Load metadata saved using [[DefaultParamsWriter.saveMetadata()]]
+   * Load metadata saved using [[DefaultParamsWriter.saveMetadataV2()]]
    *
    * @param expectedClassName  If non empty, this is checked against the loaded metadata.
    * @throws IllegalArgumentException if expectedClassName is specified and does not match metadata
    */
-  def loadMetadata(path: String, sc: SparkContext, expectedClassName: String = ""): Metadata = {
+  @deprecated("use loadMetadataV2", "4.0.0")
+  def loadMetadata(path: String, sc: SparkContext, expectedClassName: String = ""): Metadata =
+    loadMetadataV2(path, SparkSession.builder().sparkContext(sc).getOrCreate(), expectedClassName)
+
+  /**
+   * Load metadata saved using [[DefaultParamsWriter.saveMetadataV2()]]
+   *
+   * @param expectedClassName  If non empty, this is checked against the loaded metadata.
+   * @throws IllegalArgumentException if expectedClassName is specified and does not match metadata
+   */
+  def loadMetadataV2(
```

Comment:
Removing `loadMetadata` will break all 3-rd ml implementations reusing ML utils, so I leave it alone even though it is `private[ml]`

Overload with name `loadMetadata` causes compilation errors:
```
[error] /Users/ruifeng.zheng/Dev/spark/mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala:506:20: in object DefaultParamsReader, multiple overloaded alternatives of method loadMetadata define default arguments.
[error] private[ml] object DefaultParamsReader {
[error]                    ^
```

so name it `loadMetadataV2`

Reply:
Why don't you just have overwritten one instead of having `V2`?

Reply:
I tried but the compilation fails with `in object DefaultParamsReader, multiple overloaded alternatives of method loadMetadata define default arguments`

```
[error] /Users/ruifeng.zheng/Dev/spark/mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala:506:20: in object DefaultParamsReader, multiple overloaded alternatives of method loadMetadata define default arguments.
[error] private[ml] object DefaultParamsReader {
[error]     
```

Reply:
hmm, probably I can try `sc: Either[SparkContext, SparkSession]`

Reply:
This should work:


```scala
  def loadMetadata(path: String, sc: SparkContext, expectedClassName: String): Metadata = { ... }

  def loadMetadata(path: String, sc: SparkContext): Metadata = loadMetadata(path, sc, "")

  def loadMetadata(
      path: String,
      spark: SparkSession,
      expectedClassName: String): Metadata = { ... }

  def loadMetadata(
      path: String,
      spark: SparkSession): Metadata = { loadMetadata(path, spark, "") }
```

It's because of functions disallowed to be overwritten with the default values. It will also not break binary compatibility too.

Reply:
yeah, I am trying this way